routers/api/v1/repo/branch.go

```diff
@@ -173,11 +173,49 @@ func CreateBranch(ctx *context.APIContext) {
 		return
 	}
 
-	if len(opt.OldBranchName) == 0 {
-		opt.OldBranchName = ctx.Repo.Repository.DefaultBranch
+	var oldCommit *git.Commit
+	var err error
+
+	if len(opt.OldRefName) > 0 {
+		if ctx.Repo.GitRepo.IsBranchExist(opt.OldRefName) {
```

Comment:
Here and below, I think we shouldn't do the "guessing" anymore.

Is it possible to let callers provide accurate ref names? eg: `refs/heads/BRANC_NAME` and `refs/tags/TAG_NAME` etc.


zeripath's comment has a real example: 

> This (guessing) would be technically breaking because `refs/heads/deadbeef` used to refer to the ref `refs/heads/refs/heads/deadbeef`

Reply:
It looks better this way

Reply:
That way looks better

Reply:
Yes! 

If you're calling some field a ref, it had better be expecting a ref and it should not be guessing. 

This field should expect that branches provided to it are already in the form refs/heads/BRANCH and not BRANCH.